pkg/controller/servicebindingrequest/servicebinder.go

```diff
@@ -246,6 +251,77 @@ func (b *ServiceBinder) onError(
 	return RequeueOnNotFound(err, requeueAfter)
 }
 
+// isApplicationSelectorEmpty returns true if applicationSelector is not declared in
+// the Service Binding Request.
+func isApplicationSelectorEmpty(
+	application v1alpha1.ApplicationSelector,
+) bool {
+	var emptyApplication v1alpha1.ApplicationSelector
+	if application == emptyApplication ||
+		application.ResourceRef == "" &&
+			application.LabelSelector.MatchLabels == nil {
+		return true
+	}
+	return false
+}
+
+// handleApplicationSelectorEmpty handles scenario when applicationSelector is not declared in
+// the Service Binding Request.
+func (b *ServiceBinder) handleApplicationSelectorEmpty() (reconcile.Result, error) {
```

Comment:
This method is exactly the same as `handleApplicationNotFound`, with the only difference being what is returned at the end in case of success: either `Done()` as in the case of `handleApplicationSelectorEmpty` or `RequeueOnNotFound()` as in the case of `handleApplicationNotFound`.

I believe `handleApplicationNotFound` can consume the `handleApplicationSelectorEmpty` method and then deciding whether the work item should be retried:

```go
func (b *ServiceBinder) handleApplicationNotFound() (reconcile.Result, error) {
    // it is safe to ignore the first argument since a new result will be generated based 
    // on the propagated error
    _, err := b.handleApplicationSelectorEmpty();
   return RequeueOnNotFound(err)
}
```

Reply:
[1] method `handleApplicationSelectorEmpty()` and `handleApplicationNotFound()` are not exactly similar. They differ in two respect:

- reason and message of `InjectionReady` condition type is different
- what is returned at the end - Done() or RequeueOnNotFound()

[2] Based on your suggestion, If I make `handleApplicationNotFound` consume the `handleApplicationSelectorEmpty` method, then
I would have to remove these 3 lines -> https://github.com/redhat-developer/service-binding-operator/pull/498/files#diff-fae660884df4e35af60a04c5ddf930daR337-R339
Right?
that would mean that handleApplicationSelectorEmpty is called AFTER we're done searching the application -> https://github.com/redhat-developer/service-binding-operator/pull/498/files#diff-fae660884df4e35af60a04c5ddf930daR340

If the user hasn't specified any application in SBR, why should there be any search for an application at all?

Reply:
To make it clearer: I wouldn't change the existing flow so I would keep the check you've mentioned in [2] before calling `Bind`.

On the other hand, those methods are almost identical with the exception of those two values mentioned in [1] that can (and should) be parameterized in the base function itself, regardless if it is `handleApplicationSelectorEmpty` or another method you choose to extract the base function.

I think this isn't a biggie for now if we don't want to pursue this, only seemed a good improvement and easy to tackle now :)

Reply:
I get your point now. I will make changes.

Reply:
@isutton Is this what you had in mind? -> bd70eaa

Reply:
Pretty much 👍 

I believe the whole error handling could be simplified even further but for the scope of this PR naming the method handling application errors sensibly should be as far as it goes, so I propose to call the method `handleApplicationError` instead of `handleApplication` as last small change request 😅 What do you think?

Reply:
@isutton It's not really an `Error` scenario, is it?
If something is not found, it's not actually an error scenario.

Reply:
Thinking again, maybe it's okay to rename it `handleApplicationError` for now since we already have `ErrEmptyApplicationSelector` and `ErrApplicationNotFound`.

Reply:
DOne 99272e7